Crash of  all workflows (DuplicateSubscriberIdentifier)

Dear Aiida Users,

I got the following error message that lead to the crash of all running jobs/workflows/calculations during the weekend.

kiwipy.communications.DuplicateSubscriberIdentifier: Broadcast identifier 

I am usually running with several daemon threads. Any idea how I could avoid such problems in the future? 

(Running aiida 1.1.1 at the moment)



```
  File "/path/to/python/lib/python3.7/site-packages/aiida/manage/external/rmq.py", line 187, in _continue
    result = yield super()._continue(communicator, pid, nowait, tag)
  File "/path/to/python/lib/python3.7/site-packages/tornado/gen.py", line 1055, in run
    value = future.result()
  File "/path/to/python/lib/python3.7/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/path/to/python/lib/python3.7/site-packages/tornado/gen.py", line 307, in wrapper
    yielded = next(result)
  File "/path/to/python/lib/python3.7/site-packages/plumpy/process_comms.py", line 547, in _continue
    proc = saved_state.unbundle(self._load_context)
  File "/path/to/python/lib/python3.7/site-packages/plumpy/persistence.py", line 51, in unbundle
    return Savable.load(self, load_context)
  File "/path/to/python/lib/python3.7/site-packages/plumpy/persistence.py", line 447, in load
    return load_cls.recreate_from(saved_state, load_context)
  File "/path/to/python/lib/python3.7/site-packages/plumpy/processes.py", line 236, in recreate_from
    base.call_with_super_check(process.init)
  File "/path/to/python/lib/python3.7/site-packages/plumpy/base/utils.py", line 29, in call_with_super_check
    fn(*args, **kwargs)
  File "/path/to/python/lib/python3.7/site-packages/aiida/engine/processes/process.py", line 126, in init
    super().init()
  File "/path/to/python/lib/python3.7/site-packages/plumpy/base/utils.py", line 16, in new_fn
    fn(self, *args, **kwargs)
  File "/path/to/python/lib/python3.7/site-packages/plumpy/processes.py", line 295, in init
    self.broadcast_receive, identifier=str(self.pid))
  File "/path/to/python/lib/python3.7/site-packages/kiwipy/rmq/communicator.py", line 592, in add_broadcast_subscriber
    return self._run_task(coro)
  File "/path/to/python/lib/python3.7/site-packages/kiwipy/rmq/communicator.py", line 677, in _run_task
    return self.tornado_to_kiwi_future(self._create_task(coro)).result(timeout=self.TASK_TIMEOUT)
  File "/path/to/python/lib/python3.7/concurrent/futures/_base.py", line 435, in result
    return self.__get_result()
  File "/path/to/python/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/path/to/python/lib/python3.7/site-packages/kiwipy/rmq/communicator.py", line 646, in done
    result = done_future.result()
  File "/path/to/python/lib/python3.7/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/path/to/python/lib/python3.7/site-packages/kiwipy/futures.py", line 54, in capture_exceptions
    yield
  File "/path/to/python/lib/python3.7/site-packages/kiwipy/rmq/utils.py", line 146, in run_task
    future.set_result((yield coro()))
  File "/path/to/python/lib/python3.7/site-packages/tornado/gen.py", line 1055, in run
    value = future.result()
  File "/path/to/python/lib/python3.7/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/path/to/python/lib/python3.7/site-packages/tornado/gen.py", line 1063, in run
    yielded = self.gen.throw(*exc_info)
  File "/path/to/python/lib/python3.7/site-packages/kiwipy/rmq/communicator.py", line 393, in add_broadcast_subscriber
    identifier = yield self._message_subscriber.add_broadcast_subscriber(subscriber, identifier)
  File "/path/to/python/lib/python3.7/site-packages/tornado/gen.py", line 1055, in run
    value = future.result()
  File "/path/to/python/lib/python3.7/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/path/to/python/lib/python3.7/site-packages/tornado/gen.py", line 307, in wrapper
    yielded = next(result)
  File "/path/to/python/lib/python3.7/site-packages/kiwipy/rmq/communicator.py", line 146, in add_broadcast_subscriber
    raise kiwipy.DuplicateSubscriberIdentifier("Broadcast identifier '{}'".format(identifier))
  kiwipy.communications.DuplicateSubscriberIdentifier: Broadcast identifier '157299'

```

My pleasure :)

Wow, so looking at these logs it looks like the exact scenario I described is playing out, e.g.:
```
=ERROR REPORT==== 16-May-2020::23:30:20 ===
closing AMQP connection <0.23134.3> (127.0.0.1:44066 -> 127.0.0.1:5672):
missed heartbeats from client, timeout: 60s
```
Here, AiiDA has missed (presumably) two heartbeats (the default) and RabbitMQ has assumed that client to be dead.  Then, 23s later, the automatic reconnection mechanism kicks in and AiiDA reconnects:
```
=INFO REPORT==== 16-May-2020::23:30:46 ===
accepting AMQP connection <0.14036.4> (127.0.0.1:58650 -> 127.0.0.1:5672)
```
This all causes some hubbub in the daemon logs, but tellingly soon after AiiDA gets delivered the same task it was working on before it got disconnected and that causes this exception
```
05/16/2020 11:40:44 AM <15123> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [ERROR] failed to recreate the process instance in order to continue it.
```

I'm surprised because the library that communicates with RabbitMQ uses a separate thread, partly so that it can keep responding to heartbeats while other things go on.  Now, because this is python, we don't have true multithreading [1] and it's possible that your workload has blocked the interpreter for over two minutes (i.e. 2 heartbeats).

So, can I ask, what kind of workload is in your workchains?  Are you doing any heavy in-Python computation, heavy `numpy` tasks?  Or, perhaps, calling a subprocess and blocking waiting on that?  Basically anything you think could block a thread for a long time?

[1] https://realpython.com/python-gil/

Hi All, just to say that I have also experienced this today:

```
   raise self._exception
  File "/home/bonan/miniconda3/envs/aiida-1.0-py3/lib/python3.7/site-packages/kiwipy/rmq/communicator.py", line 646, in done
    result = done_future.result()
  File "/home/bonan/miniconda3/envs/aiida-1.0-py3/lib/python3.7/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/home/bonan/miniconda3/envs/aiida-1.0-py3/lib/python3.7/site-packages/kiwipy/futures.py", line 54, in capture_exceptions
    yield
  File "/home/bonan/miniconda3/envs/aiida-1.0-py3/lib/python3.7/site-packages/kiwipy/rmq/utils.py", line 146, in run_task
    future.set_result((yield coro()))
  File "/home/bonan/miniconda3/envs/aiida-1.0-py3/lib/python3.7/site-packages/tornado/gen.py", line 1055, in run
    value = future.result()
  File "/home/bonan/miniconda3/envs/aiida-1.0-py3/lib/python3.7/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/home/bonan/miniconda3/envs/aiida-1.0-py3/lib/python3.7/site-packages/tornado/gen.py", line 1063, in run
    yielded = self.gen.throw(*exc_info)
  File "/home/bonan/miniconda3/envs/aiida-1.0-py3/lib/python3.7/site-packages/kiwipy/rmq/communicator.py", line 393, in add_broadcast_subscriber
    identifier = yield self._message_subscriber.add_broadcast_subscriber(subscriber, identifier)
  File "/home/bonan/miniconda3/envs/aiida-1.0-py3/lib/python3.7/site-packages/tornado/gen.py", line 1055, in run
    value = future.result()
  File "/home/bonan/miniconda3/envs/aiida-1.0-py3/lib/python3.7/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/home/bonan/miniconda3/envs/aiida-1.0-py3/lib/python3.7/site-packages/tornado/gen.py", line 307, in wrapper
    yielded = next(result)
  File "/home/bonan/miniconda3/envs/aiida-1.0-py3/lib/python3.7/site-packages/kiwipy/rmq/communicator.py", line 146, in add_broadcast_subscriber
    raise kiwipy.DuplicateSubscriberIdentifier("Broadcast identifier '{}'".format(identifier))
kiwipy.communications.DuplicateSubscriberIdentifier: Broadcast identifier '312454'
10/08/2020 07:08:41 PM <1237> kiwipy.rmq.tasks: [ERROR] The task excepted
```

It was when I had 900 processes running with 10 workers, mostly VASP CalcJobs and WorkChains. I recall doing something very memory intensive at that time (a mongodbrestore operation....) and then notice some high usage of the python processes as well (the problem escalating). 

I have been running around 600 processes previsouly without any problem. Maybe I am pushing it too hard.....

I am running AiiDA 1.3.0 on WSL Ubuntu 18.04, the RabbitMQ service is installed on the windows side.

--- a/aiida/manage/external/rmq.py
+++ b/aiida/manage/external/rmq.py
@@ -14,7 +14,7 @@ from collections.abc import Mapping
 import logging
 import traceback
 
-from kiwipy import Future, communications
+from kiwipy import DuplicateSubscriberIdentifier, Future, communications
 import pamqp.encode
 import plumpy
 
@@ -212,6 +212,33 @@ class ProcessLauncher(plumpy.ProcessLauncher):
             message = 'the class of the process could not be imported.'
             self.handle_continue_exception(node, exception, message)
             raise
+        except DuplicateSubscriberIdentifier:
+            # This happens when the current worker has already subscribed itself with this process identifier. The call
+            # to ``_continue`` will call ``Process.init`` which will add RPC and broadcast subscribers. ``kiwipy`` and
+            # ``aiormq`` further down keep track of processes that are already subscribed and if subscribed again, a
+            # ``DuplicateSubscriberIdentifier`` is raised. Possible reasons for the worker receiving a process task that
+            # it already has, include:
+            #
+            #  1. The user mistakenly recreates the corresponding task, thinking the original task was lost.
+            #  2. RabbitMQ requeues the task because the daemon worker lost its connection or did not respond to the
+            #     heartbeat in time, and the task is sent to the same worker once it regains connection.
+            #
+            # Here we assume that the existence of another subscriber indicates that the process is still being run by
+            # this worker. We thus ignore the request to have the worker take it on again and acknowledge the current
+            # task (`return False`). If our assumption was wrong and the original task was no longer being worked on,
+            # the user can resubmit the task once the list of subscribers of the process has been cleared. Note: In the
+            # second case we are deleting the *original* task, and once the worker finishes running the process there
+            # won't be a task in RabbitMQ to acknowledge anymore. This, however, is silently ignored.
+            #
+            # Note: the exception is raised by ``kiwipy`` based on an internal cache it and ``aiormq`` keep of the
+            # current subscribers. This means that this will only occur when the tasks is resent to the *same* daemon
+            # worker. If another worker were to receive it, no exception would be raised as the check is client and not
+            # server based.
+            LOGGER.exception(
+                'A subscriber with the process id<%d> already exists, which most likely means this worker is already '
+                'working on it and this task was sent as a duplicate by mistake. Deleting the task now.', pid
+            )
+            return False
         except asyncio.CancelledError:  # pylint: disable=try-except-raise
             # note this is only required in python<=3.7,
             # where asyncio.CancelledError inherits from Exception

--- a/aiida/manage/tests/pytest_fixtures.py
+++ b/aiida/manage/tests/pytest_fixtures.py
@@ -241,7 +241,7 @@ def aiida_local_code_factory(aiida_localhost):
     return get_code
 
 
-@pytest.fixture(scope='session')
+@pytest.fixture()
 def daemon_client(aiida_profile):
     """Return a daemon client for the configured test profile for the test session.
 

--- a/tests/manage/external/test_rmq.py
+++ b/tests/manage/external/test_rmq.py
@@ -8,10 +8,14 @@
 # For further information please visit http://www.aiida.net               #
 ###########################################################################
 """Tests for the `aiida.manage.external.rmq` module."""
+import pathlib
+
 from kiwipy.rmq import RmqThreadCommunicator
 import pytest
 
+from aiida.engine.processes import ProcessState, control
 from aiida.manage.external import rmq
+from aiida.orm import Int
 
 
 @pytest.mark.parametrize(('args', 'kwargs', 'expected'), (
@@ -57,3 +61,51 @@ def test_add_rpc_subscriber(communicator):
 def test_add_broadcast_subscriber(communicator):
     """Test ``add_broadcast_subscriber``."""
     communicator.add_broadcast_subscriber(None)
+
+
+@pytest.mark.requires_rmq
+def test_duplicate_subscriber_identifier(aiida_local_code_factory, started_daemon_client, submit_and_await):
+    """Test that a ``DuplicateSubscriberError`` in ``ProcessLauncher._continue`` does not except the process.
+
+    It is possible that when a daemon worker tries to continue a process, that a ``kiwipy.DuplicateSubscriberError`` is
+    raised, which means that it already subscribed itself to be running that process.
+    This can occur for at least two reasons:
+
+        * The user manually recreated the process task, mistakenly thinking it had been lost
+        * RabbitMQ requeues the task because the daemon worker lost its connection or did not respond to the
+          heartbeat in time, and the task is sent to the same worker once it regains connection.
+
+    In both cases, the actual task is still actually being run and we should not let this exception except the entire
+    process. Unfortunately, these duplicate tasks still happen quite a lot when the daemon workers are under heavy load
+    and we don't want a bunch of processes to be excepted because of this.
+
+    In most cases, just ignoring the task wil be the best solution. In the end, the original task is likely to complete.
+    If it turns out that the task actually got lost and the process is stuck, the user can find this error message in
+    the logs and manually recreate the task, using for example ``verdi devel revive``.
+
+    Note that this exception is only raised within a single worker, i.e., if another worker subscribes to the same
+    process, that should not incur this inception and that is not what we are testing here. This test should therefore
+    be ran with a single daemon worker.
+    """
+    code = aiida_local_code_factory(entry_point='core.arithmetic.add', executable='/bin/bash')
+
+    builder = code.get_builder()
+    builder.x = Int(1)
+    builder.y = Int(1)
+    builder.metadata.options.sleep = 2  # Add a sleep to give time to send duplicate task before it finishing
+
+    # Submit the process to the daemon and wait for it to be picked up (signalled by it going in waiting state).
+    node = submit_and_await(builder, ProcessState.WAITING)
+
+    # Recreate process task causing the daemon to pick it up again and incurring the ``DuplicateSubscriberError``
+    control.revive_processes([node], wait=True)
+
+    # Wait for the original node to be finished
+    submit_and_await(node, ProcessState.FINISHED)
+
+    # The original node should now have finished normally and not excepted
+    assert node.is_finished_ok, (node.process_state, node.exit_status)
+
+    # Verify that the receiving of the duplicate task was logged by the daemon
+    daemon_log = pathlib.Path(started_daemon_client.daemon_log_file).read_text(encoding='utf-8')
+    assert f'Error: A subscriber with the process id<{node.pk}> already exists' in daemon_log

